File: golang/heap/kth_largest_element_array.go
package main

import (
	"math/rand"
)

func findKthLargest(nums []int, k int) int {
	return quickSelect(nums, 0, len(nums)-1, len(nums)-k)
}

func quickSelect(nums []int, left, right, k int) int {
	if left == right {
		return nums[left]
	}
	// rand.Seed(time.Now().UnixNano())
	pivotIndex := left + rand.Intn(right-left+1)
	pivotIndex = partition(nums, left, right, pivotIndex)

	if k == pivotIndex {
		return nums[k]
	} else if k < pivotIndex {
		return quickSelect(nums, left, pivotIndex-1, k)
	} else {
		return quickSelect(nums, pivotIndex+1, right, k)
	}

	return -1
}

func partition(nums []int, left, right, pivotIndex int) int {
	pivot := nums[pivotIndex]
	nums[pivotIndex], nums[right] = nums[right], nums[pivotIndex]

	storeIndex := left
	for i := left; i < right; i++ {
		if nums[i] < pivot {
			nums[storeIndex], nums[i] = nums[i], nums[storeIndex]
			storeIndex++
		}
	}

	nums[right], nums[storeIndex] = nums[storeIndex], nums[right]
	return storeIndex
}

File: python/heap/kth_largest_element_array.py
from random import randint

from typing import List


class Solution:
    def findKthLarges(self, nums: List[int], k: int) -> int:
        return self.quickSelect(nums, 0, len(nums)-1 , len(nums) - k)

    def quickSelect(self, nums: List[int], left:  int, right: int, k: int) -> int:
        if left == right:
            return nums[left]

        pivotIndex = left + randint(0,right - left + 1)
        pivotIndex = self.partition(nums, left, right, pivotIndex)

        if k == pivotIndex:
            return nums[k]
        elif k < pivotIndex:
            return self.quickSelect(nums, left, pivotIndex - 1, k)
        else:
            return self.quickSelect(nums, pivotIndex + 1, right, k)

        return -1

    def partition(self, nums: List[int], left: int, right: int, pivotIndex: int) -> int:
        pivot = nums[pivotIndex]

        nums[pivotIndex], nums[right] = nums[right], nums[pivotIndex]

        storeIndex = left


        for i in range(left, right):
            if nums[i] < pivot:
                nums[storeIndex], nums[i] = nums[i], nums[storeIndex]
                storeIndex += 1

        nums[right], nums[storeIndex] = nums[storeIndex], nums[right]

        return storeIndex
